engine/execution/ingestion/core.go

```diff
@@ -0,0 +1,461 @@
+
+	// running worker in the background to consume
+	// processables blocks which are throttled,
+	// and forward them to the block queue for processing
+	e.unit.Launch(func() {
+		e.log.Info().Msgf("starting worker to consume throttled blocks")
+		err := e.forwardProcessableToHandler(processables)
+		if err != nil {
+			e.log.Fatal().Err(err).Msg("fail to process block")
+		}
+	})
+
+	e.log.Info().Msg("initializing throttle engine")
+
+	err := e.throttle.Init(processables)
+	if err != nil {
+		e.log.Fatal().Err(err).Msg("fail to initialize throttle engine")
+	}
+
+	e.log.Info().Msgf("throttle engine initialized")
+}
+
+func (e *Core) forwardProcessableToHandler(
+	processables <-chan flow.Identifier,
+) error {
+	for blockID := range processables {
+		err := e.onProcessableBlock(blockID)
+		if err != nil {
+			return fmt.Errorf("could not process block: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func (e *Core) onProcessableBlock(blockID flow.Identifier) error {
+	header, err := e.headers.ByBlockID(blockID)
+	if err != nil {
+		return fmt.Errorf("could not get block: %w", err)
+	}
+
+	// skip if stopControl tells to skip
+	if !e.stopControl.ShouldExecuteBlock(header) {
+		return nil
+	}
+
+	executed, err := e.execState.IsBlockExecuted(header.Height, blockID)
+	if err != nil {
+		return fmt.Errorf("could not check whether block %v is executed: %w", blockID, err)
+	}
+
+	if executed {
+		e.log.Debug().Msg("block has been executed already")
+		return nil
+	}
+
+	block, err := e.blocks.ByID(blockID)
+	if err != nil {
+		return fmt.Errorf("failed to get block %s: %w", blockID, err)
+	}
+
+	missingColls, executables, err := e.enqueuBlock(block, blockID)
+	if err != nil {
+		return fmt.Errorf("failed to enqueue block %v: %w", blockID, err)
+	}
+
+	e.executeConcurrently(executables)
+
+	err = e.fetch(missingColls)
+	if err != nil {
+		return fmt.Errorf("failed to fetch missing collections: %w", err)
+	}
+
+	return nil
+}
+
+func (e *Core) enqueuBlock(block *flow.Block, blockID flow.Identifier) (
+	[]*block_queue.MissingCollection, []*entity.ExecutableBlock, error) {
+	lg := e.log.With().
+		Hex("block_id", blockID[:]).
+		Uint64("height", block.Header.Height).
+		Logger()
+
+	lg.Info().Msg("handling new block")
+
+	parentCommitment, err := e.execState.StateCommitmentByBlockID(block.Header.ParentID)
+
+	if err == nil {
+		// the parent block is an executed block.
+		missingColls, executables, err := e.blockQueue.HandleBlock(block, &parentCommitment)
+		if err != nil {
+			return nil, nil, fmt.Errorf("unexpected error while adding block to block queue: %w", err)
+		}
+
+		lg.Info().Bool("parent_is_executed", true).
+			Int("missing_col", len(missingColls)).
+			Int("executables", len(executables)).
+			Msgf("block is enqueued")
+
+		return missingColls, executables, nil
+	}
+
+	// handle exception
+	if !errors.Is(err, storage.ErrNotFound) {
+		return nil, nil, fmt.Errorf("failed to get parent state commitment for block %v: %w", block.Header.ParentID, err)
```

Comment:
should this list the enqueued block?
```suggestion
		return nil, nil, fmt.Errorf("failed to get parent state commitment for block %v: %w", blockID, err)
```